cmake/modules/boards.cmake

```diff
@@ -95,41 +109,119 @@ Hints:
+endif()
+
+# Prepare list boards command.
+# This command is used for locating the board dir as well as printing all boards
+# in the system in the following cases:
+# - User specifies an invalid BOARD
+# - User invokes '<build-command> boards' target
+list(TRANSFORM ARCH_ROOT PREPEND "--arch-root=" OUTPUT_VARIABLE arch_root_args)
+list(TRANSFORM BOARD_ROOT PREPEND "--board-root=" OUTPUT_VARIABLE board_root_args)
+
+set(list_boards_commands
+    COMMAND ${PYTHON_EXECUTABLE} ${ZEPHYR_BASE}/scripts/list_boards.py
+            ${arch_root_args} ${board_root_args} --arch-root=${ZEPHYR_BASE}
+)
+
+if(NOT BOARD_DIR)
+  if(BOARD_ALIAS)
+    execute_process(${list_boards_commands} --board=${BOARD_ALIAS} --format={name}\;{dir}\;{hwm}
+                    OUTPUT_VARIABLE ret_board
+                    ERROR_VARIABLE err_board
+                    RESULT_VARIABLE ret_val
+    )
+    list(GET ret_board 1 BOARD_HIDDEN_DIR)
+    string(STRIP "${BOARD_HIDDEN_DIR}" BOARD_HIDDEN_DIR)
+    set(BOARD_HIDDEN_DIR ${BOARD_HIDDEN_DIR} CACHE PATH "Path to a folder." FORCE)
 
-  # NB: find_path will return immediately if the output variable is
-  # already set
-  if (BOARD_ALIAS)
-    find_path(BOARD_HIDDEN_DIR
-      NAMES ${BOARD_ALIAS}_defconfig
-      PATHS ${root}/boards/*/*
-      NO_DEFAULT_PATH
-      )
     if(BOARD_HIDDEN_DIR)
       message("Board alias ${BOARD_ALIAS} is hiding the real board of same name")
     endif()
   endif()
-  if(BOARD_DIR AND NOT EXISTS ${BOARD_DIR}/${BOARD}_defconfig)
-    message(WARNING "BOARD_DIR: ${BOARD_DIR} has been moved or deleted. "
-                    "Trying to find new location."
-    )
-    set(BOARD_DIR BOARD_DIR-NOTFOUND CACHE PATH "Path to a file." FORCE)
+
+  execute_process(${list_boards_commands} --board=${BOARD}
+    --cmakeformat={NAME}\;{DIR}\;{HWM}\;{REVISION_FORMAT}\;{REVISION_DEFAULT}\;{REVISIONS}\;{CPUSET_DEFAULT}\;{CPUSETS}
+                  OUTPUT_VARIABLE ret_board
+                  ERROR_VARIABLE err_board
+                  RESULT_VARIABLE ret_val
+  )
+  if(ret_val)
+    message(FATAL_ERROR "Error finding board: ${BOARD}\nError message: ${err_board}")
   endif()
-  find_path(BOARD_DIR
-    NAMES ${BOARD}_defconfig
-    PATHS ${root}/boards/*/*
-    NO_DEFAULT_PATH
-    )
-  if(BOARD_DIR AND NOT (${root} STREQUAL ${ZEPHYR_BASE}))
-    set(USING_OUT_OF_TREE_BOARD 1)
+  string(STRIP "${ret_board}" ret_board)
+  cmake_parse_arguments(BOARD "" "NAME;DIR;HWM;REVISION_FORMAT;REVISION_DEFAULT;CPUSET_DEFAULT" "REVISIONS;CPUSETS" ${ret_board})
+  set(BOARD_DIR ${BOARD_DIR} CACHE PATH "Board directory for board (${BOARD})" FORCE)
+
+  # Create two CMake variables identifying the hw model.
+  # CMake variable: HWM=[v1,v2]
+  # CMake variable: HWMv1=True, when HWMv1 is in use.
+  # CMake variable: HWMv2=True, when HWMv2 is in use.
+  set(HWM       ${BOARD_HWM} CACHE INTERNAL "Zephyr hardware model version")
+  set(HWM${HWM} True   CACHE INTERNAL "Zephyr hardware model")
```

Comment:
What if someone configured a project in HWMv1 mode, updated their board to HWMv2 then just built the existing configured project? HWM would be updated but HWMv1 and HWMv2 would both be set to true?

Reply:
that's a good point.

There will be an automatic CMake re-run due to changes of CMake and Kconfig files during the update, but I think we should fail gracefully with information to the user that a pristine build is required, and why.

To be addressed before ready for merge.